File: GithubUser/GithubUser/Manager/ApiClient.swift
//
//  ApiClient.swift
//  GithubUser
//
//  Created by Lam Le on 15/06/2022.
//

import Foundation
import Alamofire
import RxSwift
import RxCocoa

enum NetworkError: Swift.Error {
    case invalidUrl
    case disconnect
    case timeout
    case cancel
    case noResponse
    case error(Error)
}

typealias Completion<Value> = (Result<Value, NetworkError>) -> Void
typealias JSON = [String: Any]

protocol Cancellable {
    func cancel()
}

protocol ApiClientProtocol: ReactiveCompatible {
    func makeCall(_ router: Router, completion: @escaping Completion<Data>) -> Request?
}

final class ApiClient: ApiClientProtocol {

    private let manager: Session
    private var request: Request?

    static let shared = ApiClient()

    private init() {
        // Configuration
        let configuration: URLSessionConfiguration = .default
        configuration.timeoutIntervalForRequest = 30
        configuration.timeoutIntervalForResource = 30
        configuration.waitsForConnectivity = true

        manager = Session(
            configuration: configuration
        )
    }

    // MARK: - Old implementation
    @discardableResult
    func makeCall(
        _ router: Router,
        completion: @escaping Completion<Data>) -> Request? {
            guard NetworkReachabilityManager()?.isReachable == true else {
                completion(.failure(.disconnect))
                return nil
            }
            self.request = manager.request(router).response { response in
                guard NetworkReachabilityManager()?.isReachable == true else {
                    completion(.failure(.disconnect))
                    return
                }
                let statusCode = response.response?.statusCode
                if statusCode == NSURLErrorTimedOut {
                    completion(.failure(.timeout))
                    return
                }
                if statusCode == NSURLErrorCancelled {
                    completion(.failure(.cancel))
                    return
                }
                if let error = response.error {
                    completion(.failure(.error(error)))
                    return
                }
                guard let data = response.data else {
                    completion(.failure(.noResponse))
                    return
                }
                completion(.success(data))
            }
            return request
        }
}

// MARK:
extension ApiClient: Cancellable {

    func cancel() {
        request?.cancel()
    }
}

// MARK: - ReactiveCompatible
extension ApiClient: ReactiveCompatible {}

// MAKR: - ApiClientService
protocol ApiClientService {
    func makeCall(_ router:  Router) -> Observable<Result<Data, NetworkError>>
}
// MARK: - Reactive

extension Reactive: ApiClientService where Base: ApiClient {

    func makeCall(_ router:  Router) -> Observable<Result<Data, NetworkError>> {
        return Observable.create { observer in
            base.makeCall(router) { result in
                observer.onNext(result)
            }
            return Disposables.create {
                base.cancel()
            }
        }
    }
}

// MARK: - Config
extension ApiClient {

    struct Config {
        static let timeOut: TimeInterval = 30
    }
}

File: fl_github_users/lib/api_client/api_client.dart
import 'package:http/http.dart' as http;

enum Method {
  get
}

abstract class Router {
  String get uri;
  Method get method;
  Map<String, String>? get headers;
}

abstract class ApiClient {
  Future<Result<String>> fetch(Router router);
}

class ApiClientImpl implements ApiClient {

  final http.Client _client = http.Client();

  @override
  Future<Result<String>> fetch(Router router) async {
    var response = await _client.get(Uri.parse(router.uri), headers: router.headers);
    try {
      switch (response.statusCode) {
        case 200:
          return Result.success(response.body);
        case -1005:
          throw Result.failure(NetworkError.disconnect());
        case -1001:
          throw Result.failure(NetworkError.timeout());
        case -999:
          throw Result.failure(NetworkError.cancel());
        default:
          throw Result.failure(NetworkError.unknown(response.statusCode));
      }
    } catch (e) {
      throw Result.failure(NetworkError.unknown(1000));
    }
   }
}

class Result<T> implements Exception {
  final T? value;
  final Exception? error;
  const Result({this.value, this.error});
  factory Result.success(T value) {
    return Result(value: value, error: null);
  }
  factory Result.failure(Exception error) {
    return Result(value: null, error: error);
  }
}

class NetworkError implements Exception {
  int code;
  String message;

  NetworkError(this.code, this.message);

  factory NetworkError.disconnect() => NetworkError(-1005, "network error");
  factory NetworkError.timeout() => NetworkError(-1001, "time out");
  factory NetworkError.unknown(int code) => NetworkError(code, "Unknown");
  factory NetworkError.cancel() => NetworkError(-999, "Cancel request");
}
